[ADD][TAK][3446822] add presale module

NEDJIMAbelgacem opened the pull request:
-  PR containing the code for task [SH Development - bene](https://www.odoo.com/web#id=3446822&cids=1&menu_id=4720&action=333&active_id=5667&model=project.task&view_type=form)
- Functional analysis: [Here](https://www.odoo.com/web#id=2853025&cids=1&model=project.task&view_type=form)
- Technical analysis: [Here](https://www.odoo.com/web#id=3446818&cids=1&menu_id=4720&action=333&active_id=5667&model=project.task&view_type=form)

NEDJIMAbelgacem committed:
[ADD][TAK][3446822] add presale module

NEDJIMAbelgacem committed:
:heart_eyes: add pre-commit configs

see https://github.com/odoo-ps/psbe-process/wiki/Development-common-practices#pre-commit

NEDJIMAbelgacem committed:
[FIX] add --exit-zero to pylint with mandatory checks hook

NEDJIMAbelgacem committed:
[IMP] apply precommit hooks to project

Simon-sila commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,136 @@
+        name: pylint with mandatory checks
+        args:
+          - --rcfile=.pylintrc-mandatory
+          - --exit-zero
```
Hey !
After testing locally, i think this should not be added,
please run
`pre-commit run --all-files`
then fix the failing checks : you have to add a readme inside the presale module and add the 'license' key to the manifest of the module.

Simon-sila commented on presale/models/presale_model.py:
```diff
@@ -0,0 +1,66 @@
+            if record.state == "new":
+                record.state = "cancelled"
+
+    def archive_confirmed_orders(self):
```
I think this function could be optimised a bit, as you are in the presale.order model already i think something like
```py
self.filtered(lambda rec : rec.state == "validated").action_archive()
```
should work and is more optimised. First because as you already have the recordset in self, filtered is better and then we take advantage of the lambda to only do the modification on some records

Simon-sila commented on presale/models/presale_model.py:
the convention is to name the file in the same way as the model (so here presale_order)

Simon-sila commented on presale/views/sale_order_views.xml:
```diff
@@ -0,0 +1,36 @@
+                    string="PreSale Order"
+                    attrs="{'readonly': True}"
+                >
+                    <!-- <field name="presale_order" attrs="{'readonly': True}"/> -->
```
Could you removed commented code if you don't use it ?

Simon-sila commented on presale/models/presale_model.py:
```diff
@@ -0,0 +1,66 @@
+    _name = "presale.order"
+    _description = "Presale order model"
+
+    def _defaultName(self):
```
Not convinced about this, to me we should keep things as simple as possible.
Like others said in the thread the other day, i don't think its really necessary, and the standard way of Odoo to deal with sequence numbers is to use ir.sequence, please us that instead.

Simon-sila commented on presale/models/presale_model.py:
```diff
@@ -0,0 +1,66 @@
+                    )
+                self.send_validation_email(record)
+
+    def action_cancel_preorder(self):
```
I think this function can be optimised a bit, please try if something like
```py
self.filtered(lambda rec : rec.state == "new").state = "cancelled"
```
works the way you want it to

Simon-sila commented on presale/models/presale_model.py:
```diff
@@ -0,0 +1,66 @@
+                    }
+                )
+                sale_order.presale_order = record.id
+                for presale_order_line in record.order_line_ids:
```
I think you should rewrite this so as to call create only once.

for example you could do 
```py
sale_order_lines = list()
for presale_order_line in record.order_line_ids:
             sale_order_lines.append({....})
self.env["sale.order.line"].create(sale_order_lines)
```

Simon-sila commented on presale/views/cron_action.xml:
we usually put cron actions, ir.sequences and email templates in a 'data' folder, could you please move this to a new 'data' folder ?

Simon-sila commented on presale/models/presale_model.py:
```diff
@@ -0,0 +1,66 @@
+    )
+    order_line_ids = fields.One2many("presale.order.line", "presale_order_id")
+
+    def send_validation_email(self, presale_order):
```
Here you can use a mail template instead of creating the body directly in python.
That way the mail template can be modified more easily !
if you want i can give you an example of mail template i made for this training

Simon-sila reviewed (commented):
Hey @NEDJIMAbelgacem !
I made some comments on your pull request, but its very good work overall

NEDJIMAbelgacem committed:
[FIX] Fix pre-commit warning

NEDJIMAbelgacem committed:
[FIX][MOV] presale module: Address SILA reviews
- fix precommit tests
- optimize presale record related functions
- rename presale.order class and file
- remove commented code
- create order lines in one go
- move cron action to data/
- use ir.sequence for presale order name
- use email template

NEDJIMAbelgacem commented:
Hi @Simon-sila 
All reviews have been addressed, let me know if anything else needs to be changed

Simon-sila commented:
perfect, excellent work ! @NEDJIMAbelgacem

Simon-sila reviewed (approved):
You can now squash and merge your commits into main !
